pf/tfbridge/ignore_changes.go

```diff
@@ -0,0 +1,301 @@
+// limitations under the License.
+
+package tfbridge
+
+import (
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/convert"
+)
+
+// Implements support for ignoreChanges property.
+//
+// There is some complexity here since Plugin Framework returns diffs in terms of Terraform paths, so the implementation
+// tries to translate TF paths to Pulumi paths to decide if they match.
+//
+// See also https://www.pulumi.com/docs/intro/concepts/resources/options/ignorechanges/
+type ignoreChanges struct {
+	// Delegate syntax parsing to resource.PropertyPath.
+	paths []resource.PropertyPath
+	ss    schemaStepper
+}
+
+func newIgnoreChanges(
+	schema *schema.PackageSpec,
+	token tokens.Token,
+	renames convert.PropertyNames,
+	rawIgnoreChanges []string,
+) (*ignoreChanges, error) {
+	var paths []resource.PropertyPath
+	for _, p := range rawIgnoreChanges {
+		pp, err := resource.ParsePropertyPath(p)
+		if err != nil {
+			return nil, err
+		}
+		paths = append(paths, pp)
+	}
+	return &ignoreChanges{
+		paths: paths,
+		ss: &namedEntitySchemaStepper{
+			schema: schema,
+			token:  token,
+			// Another complication is that Pulumi renames the properties, this uses the Renames framework
+			// to track the name tables.
+			renames: renames,
+		},
+	}, nil
+}
+
+func (ic *ignoreChanges) IsIgnored(ap *tftypes.AttributePath) bool {
+	for _, p := range ic.paths {
+		if ic.isIgnoredBy(p, ap) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ic *ignoreChanges) isIgnoredBy(pattern resource.PropertyPath, ap *tftypes.AttributePath) bool {
+	mc := &matchingContext{
+		remainingPattern: pattern,
+		schemaStepper:    ic.ss,
+	}
+	match, remain, err := tftypes.WalkAttributePath(mc, ap)
+	if err != nil {
+		return false
+	}
+	if remain != nil && len(remain.Steps()) > 0 {
+		return false
+	}
+	if m, ok := match.(*matchingContext); ok {
+		if !m.matchFailed {
+			return true
+		}
+	}
+	return false
+}
+
+type schemaStepper interface {
+	Property(resource.PropertyKey) schemaStepper
+	Element() schemaStepper
+	PropertyKey(convert.TerraformPropertyName) resource.PropertyKey
+}
+
+type typeSchemaStepper struct {
+	renames convert.PropertyNames
+	schema  *schema.PackageSpec
+	t       *schema.TypeSpec
+}
+
+var _ schemaStepper = (*typeSchemaStepper)(nil)
+
+func (p *typeSchemaStepper) Element() schemaStepper {
+	if p.t.AdditionalProperties != nil {
+		return typeStepper(p.renames, p.schema, p.t.AdditionalProperties)
+	}
+	if p.t.Items != nil {
+		return typeStepper(p.renames, p.schema, p.t.Items)
+	}
+	return nil
+}
+
+func (*typeSchemaStepper) Property(k resource.PropertyKey) schemaStepper {
+	// there are no named properties here, those are supported by namedEntitySchemaStepper
```

Comment:
I thought you could have a map with an ignored property. Is that wrong?

Reply:
Sure and it works https://github.com/pulumi/pulumi-terraform-bridge/blob/509f13f49641c069d7a7c9259fada738909e3002/pf/tfbridge/ignore_changes_test.go#L128-#L128

TF framework makes a distinction between tftypes.ElementKeyString and tftypes.AttributeName in paths. It actually helps us here because then we know not to apply name mangling to tftypes.ElementKeyString. Property func is for situations where name mangling is required (attirs).